### ISSUE TITLE
🚀 Faz 3: GEMMA Model Eğitim Pipeline Entegrasyonu

### ISSUE THREAD

**Issue by SefaGH** [SefaGH/bearish-alpha-bot#350]
### **Açıklama (Body):**

Bu görev, "Bearish Alpha Bot" projesinin GEMMA entegrasyonu yol haritasının üçüncü ve son fazını oluşturmaktadır. Bu fazda, Faz 2'de hazırlanan 82 özellikli veri setini kullanarak GEMMA modelini eğitecek mantık, projenin merkezi model eğitim script'i olan `scripts/train_all_models.py` dosyasına entegre edilecektir.

**Entegrasyonun Temel Hedefleri:**
1.  **Merkezi Konfigürasyon:** GEMMA modelinin mimarisini, eğitim parametrelerini ve başarı eşiklerini içeren bir konfigürasyon bloğunu, mevcut `config.example.yaml` ve `LiveTradingConfiguration` sistemine eklemek.
2.  **Yeni Eğitim Fonksiyonu:** Yeni 82 özellikli veri setini işleyebilen, PyTorch tabanlı GEMMA modelini eğiten, doğrulayan ve kaydeden yeni bir `train_gemma_model` fonksiyonu oluşturmak.
3.  **Pipeline Entegrasyonu:** Mevcut `main` eğitim akışını, RL ajanından sonra ve metriklerin kaydedilmesinden önce çalışacak şekilde, bu yeni fonksiyonu şartlı olarak (`ML__GEMMA__ENABLED` ortam değişkeni ile) çağırabilecek şekilde güncellemek.
4.  **Performans Takibi:** Eğitim sonuçlarını, projenin mevcut `ModelPerformanceTracker` sistemine kaydetmek.

---

### **✅ Kabul Kriterleri (Acceptance Criteria)**

- **Görev 1:** `config/config.example.yaml` dosyası, `ml` bloğu altına `gemma` adında yeni bir konfigürasyon bölümü içerecek şekilde güncellenmelidir.
- **Görev 2:** `scripts/train_all_models.py` dosyasına, GEMMA modelini eğiten, doğrulayan ve kaydeden `train_gemma_model` adında yeni bir fonksiyon eklenmelidir.
- **Görev 3:** `train_gemma_model` fonksiyonu, `ModelPerformanceTracker`'ı kullanarak eğitim metriklerini (`accuracy`, `loss`, `training_time` vb.) kaydetmelidir.
- **Görev 4:** `main` fonksiyonu, `config.ml.gemma.enabled` bayrağını kontrol ederek `train_gemma_model` fonksiyonunu, RL eğitimi adımından sonra şartlı olarak çağırmalıdır.
- **Görev 5:** Faz 2'de geliştirilen ve 82 yeni özelliği üreten mantığın, `FeatureEngineeringPipeline` içine entegre edildiği ve `GEMMA` eğitimi için bu verinin hazır olduğu varsayılmaktadır. Bu entegrasyon ayrı bir görevdir ancak bu issue'nun tamamlanabilmesi için bir ön koşuldur.

---

#### **Adım 1: Konfigürasyonu Merkezi Sisteme Ekle**

`config/config.example.yaml` dosyasını açın ve `ml` bloğunun içine `gemma` yapılandırmasını ekleyin. Bu, `GEMMA_CONFIG`'in Python içinde manuel olarak tanımlanmasını önler ve yapılandırmayı merkezi hale getirir.

```yaml
# Yerleştirme: config/config.example.yaml (ml bloğunun altına)

  gemma:
    # Ortam Değişkeni: ML__GEMMA__ENABLED
    enabled: false
    # Ortam Değişkeni: ML__GEMMA__FEATURE_SET
    feature_set: 'gemma_v1' # Faz 2'de tanımlanan özellik setinin adı
    
    architecture:
      # Ortam Değişkeni: ML__GEMMA__ARCHITECTURE__INPUT_SIZE
      input_size: 82
      hidden_size: 32
      num_layers: 2
      dropout: 0.6
      num_classes: 3  # bearish, neutral, bullish
    
    training:
      # Ortam Değişkeni: ML__GEMMA__TRAINING__EPOCHS
      epochs: 50
      batch_size: 32
      learning_rate: 0.001
      early_stopping_patience: 10
      
    thresholds:
      # Ortam Değişkeni: ML__GEMMA__THRESHOLDS__DEPLOYMENT_ACCURACY
      deployment_accuracy: 0.78
      min_samples: 1000
```

#### **Adım 2: GEMMA Eğitim Fonksiyonunu Ekle (`train_all_models.py`)**

Bu yeni fonksiyonu, dosyanın sonuna, `main` fonksiyonundan önce ekleyin. Fonksiyon, veri hazırlama, model oluşturma, eğitim döngüsü ve model kaydetme adımlarını içermelidir.

```python
# Yerleştirme: scripts/train_all_models.py (yaklaşık satır 620+)
from pathlib import Path
import shutil
import joblib

def train_gemma_model(training_data: dict, feature_engine: FeatureEngineeringPipeline, config: dict, tracker: ModelPerformanceTracker):
    """
    Train a GEMMA-based price movement prediction model.
    """
    gemma_config = config.get('gemma', {})
    if not gemma_config.get('enabled', False):
        logger.info("GEMMA training is disabled via config. Skipping.")
        return {}

    logger.info("\n" + "="*60)
    logger.info("💎 ADIM 4: GEMMA MODELİ EĞİTİLİYOR 💎")
    logger.info("="*60)
    
    # Gerekli kütüphaneleri import et (lazy-loading)
    import torch
    import torch.nn as nn
    from torch.utils.data import DataLoader, TensorDataset
    from sklearn.preprocessing import StandardScaler
    from sklearn.model_selection import train_test_split
    
    gemma_results = {}
    training_start_time = datetime.now()
    
    try:
        symbol = list(training_data.keys())[0]
        # GEMMA, en zengin veri setini kullanmak için en uzun zaman dilimini hedefler.
        timeframe = '1d' if '1d' in training_data[symbol] else ALL_TIMEFRAMES[-1]
        
        logger.info(f"Preparing data for GEMMA using symbol '{symbol}' and timeframe '{timeframe}'")
        raw_data = training_data[symbol][timeframe]

        # 1. ÖZELLİKLERİ VE ETİKETLERİ HAZIRLA
        # Faz 2'de FeatureEngineeringPipeline'e eklendiği varsayılan yeni özellik setini kullan.
        feature_set_name = gemma_config.get('feature_set', 'gemma_v1')
        logger.info(f"Extracting feature set: '{feature_set_name}'")
        features_df = feature_engine.extract_features(raw_data.copy(), feature_set=feature_set_name)
        
        # Etiketler, fiyat tahmin modelleriyle tutarlı olmalı.
        # Bu kısım, projedeki etiketleme mantığına göre uyarlanmalıdır.
        # Örnek olarak basit bir etiketleme:
        features_df['target'] = (features_df['close'].shift(-5) > features_df['close']).astype(int)
        features_df.dropna(inplace=True)

        features = features_df.drop(columns=['target'])
        labels = features_df['target']

        if features.shape[0] < gemma_config['thresholds']['min_samples']:
            logger.warning(f"⚠️ Not enough data for GEMMA training. Have {features.shape[0]}, need {gemma_config['thresholds']['min_samples']}.")
            return {'status': 'skipped', 'reason': 'insufficient_data'}

        # 2. VERİYİ ÖLÇEKLENDİR (SCALER)
        scaler = StandardScaler()
        features_scaled = scaler.fit_transform(features)
        
        scaler_path = Path('data/cache/gemma/scaler_gemma.joblib')
        scaler_path.parent.mkdir(parents=True, exist_ok=True)
        joblib.dump(scaler, scaler_path)
        logger.info(f"✅ GEMMA scaler saved to {scaler_path}")

        # 3. VERİ SETİNİ TRAIN VE VALIDATION OLARAK AYIR
        X_train, X_val, y_train, y_val = train_test_split(
            features_scaled, labels.values, test_size=0.2, random_state=42, stratify=labels
        )
        
        train_dataset = TensorDataset(torch.tensor(X_train, dtype=torch.float32), torch.tensor(y_train, dtype=torch.long))
        val_dataset = TensorDataset(torch.tensor(X_val, dtype=torch.float32), torch.tensor(y_val, dtype=torch.long))
        
        train_loader = DataLoader(train_dataset, batch_size=gemma_config['training']['batch_size'], shuffle=True)
        val_loader = DataLoader(val_dataset, batch_size=gemma_config['training']['batch_size'])

        # 4. GEMMA MODELİNİ TANIMLA (Basit bir MLP olarak)
        arch = gemma_config['architecture']
        model = nn.Sequential(
            nn.Linear(arch['input_size'], arch['hidden_size']), nn.ReLU(), nn.Dropout(arch['dropout']),
            nn.Linear(arch['hidden_size'], arch['hidden_size']), nn.ReLU(), nn.Dropout(arch['dropout']),
            nn.Linear(arch['hidden_size'], arch['num_classes'])
        )
        
        # 5. EĞİTİM DÖNGÜSÜ
        criterion = nn.CrossEntropyLoss()
        optimizer = torch.optim.Adam(model.parameters(), lr=gemma_config['training']['learning_rate'])
        
        best_accuracy = 0.0
        patience_counter = 0
        
        for epoch in range(gemma_config['training']['epochs']):
            # ... (eğitim ve validasyon döngüsü orijinal taslaktaki gibi kalabilir) ...
            # ... (içerik uzun olmasın diye çıkarıldı, geliştirici orijinal taslaktan kopyalayabilir)
            # Sadece best_accuracy güncellemesinden sonraki loglamayı düzeltelim:
            accuracy = 0.8 # Örnek
            if accuracy > best_accuracy:
                best_accuracy = accuracy
                patience_counter = 0
                model_path = Path('data/models/gemma/staging/gemma_price.pt')
                model_path.parent.mkdir(parents=True, exist_ok=True)
                model_scripted = torch.jit.script(model)
                model_scripted.save(str(model_path))
                logger.info(f"✅ New best GEMMA model saved with accuracy {best_accuracy:.4f} to {model_path}")

        training_time = (datetime.now() - training_start_time).total_seconds()
        
        # 6. SONUÇLARI KAYDET VE RAPORLA
        gemma_results = {
            'model_type': 'GEMMA',
            'final_accuracy': best_accuracy,
            'model_path': str(model_path)
        }
        
        if best_accuracy >= gemma_config['thresholds']['deployment_accuracy']:
            logger.info(f"✅ GEMMA model passed deployment threshold ({best_accuracy:.4f} >= {gemma_config['thresholds']['deployment_accuracy']}).")
            final_path = Path('data/models/gemma/final/gemma_price.pt')
            final_path.parent.mkdir(parents=True, exist_ok=True)
            shutil.copy(model_path, final_path)
            logger.info(f"✅ Model promoted to {final_path}")
        else:
            logger.warning(f"⚠️ GEMMA model accuracy ({best_accuracy:.4f}) is below deployment threshold.")

        # Performans tracker'a kaydet
        tracker.record_training(
            model_type="gemma",
            model_name=f"{symbol.replace('/', '-')}_{timeframe}",
            metrics={'accuracy': best_accuracy, 'epochs': epoch + 1},
            data_info={'samples': len(features), 'features': arch['input_size']},
            training_time=training_time
        )
        return gemma_results

    except Exception as e:
        logger.error(f"❌ GEMMA model training failed: {e}", exc_info=True)
        return {'status': 'failed', 'error': str(e)}

```

#### **Adım 3: Ana Eğitim Akışını Güncelle (`main` fonksiyonu)**

`main` fonksiyonunda, RL ajanı eğitiminden sonra, metriklerin kaydedilmesinden önce GEMMA eğitimini tetikleyecek bloğu ekleyin.

```python
# Yerleştirme: scripts/train_all_models.py -> main() (yaklaşık satır 615 civarı, RL bloğundan sonra)
    else:
        logger.error(f"RL eğitimi için gerekli olan {symbol_for_rl} sembolüne ait {RL_TRAINING_TIMEFRAME} verisi bulunamadı.")
        # ... (mevcut RL hata bloğu)

    # =========================================================================
    # 💎 ADIM 4: GEMMA MODELİ EĞİTİMİ (YENİ BLOK) 💎
    # =========================================================================
    if config.ml.gemma.enabled:
        gemma_training_results = train_gemma_model(
            training_data=training_data,
            feature_engine=feature_engine,
            config=config.ml,  # Pass the entire ML config block
            tracker=tracker
        )
        training_metrics['gemma_models'] = gemma_training_results
        logger.info(f"GEMMA Training Results: {gemma_training_results}")
    else:
        training_metrics['gemma_models'] = {'status': 'disabled'}


    # Save training metrics to files
    end_time = datetime.now()
    # ... (mevcut metrik kaydetme kodları)
```
